libp2p/Peer.cpp

```diff
@@ -45,18 +45,64 @@ Peer::Peer(Peer const& _original):
 
 bool Peer::shouldReconnect() const
 {
-    return id && endpoint && chrono::system_clock::now() > m_lastAttempted + chrono::seconds(fallbackSeconds());
+    return id && endpoint &&
+           fallbackSeconds() != numeric_limits<unsigned>::max() &&
+           chrono::system_clock::now() > m_lastAttempted + chrono::seconds(fallbackSeconds());
 }
-    
+
+bool Peer::uselessPeer() const
+{
+    if (peerType == PeerType::Required)
+        return false;
+
+    switch (m_lastHandshakeFailure)
+    {
+    case FrameDecryptionFailure:
+    case ProtocolError:
+        return true;
+    default:
+        break;
+    }
+
+    switch (m_lastDisconnect)
+    {
+    case BadProtocol:
+    case UselessPeer:
+    case IncompatibleProtocol:
+    case UnexpectedIdentity:
+    case UserReason:
```

Comment:
Not sure about `UserReason` - this could in theory be any kind of reason specific to subprotocol, including some temporary reasons, maybe it makes sense to try to reconnect

Reply:
I chose to treat UserReason as terminal because we only use it in 2 cases - if status validation fails: https://github.com/ethereum/aleth/blob/505aeada27f396dd5cb11e69d5181ca84a914186/libethereum/BlockChainSync.cpp#L190-L202
Here's where we perform the actual validation in EthereumPeer::validate: https://github.com/ethereum/aleth/blob/505aeada27f396dd5cb11e69d5181ca84a914186/libethereum/EthereumPeer.cpp#L58-L68

* If there's a bug in our network code and Session read validation fails: https://github.com/ethereum/aleth/blob/505aeada27f396dd5cb11e69d5181ca84a914186/libp2p/Session.cpp#L409-L418

You bring up a good point though, technically it can be any reason specific to a subprotocol and other clients can also choose to send it to us for reasons specific to their implementation. I think that rather than treating it as immediately critical we should use it in some sort of reconnection count threshold.

Reply:
@gumb0 I've decided to change `UserReason `-> `UselessPeer `when status message validation fails for a peer (since if that happens the peer is effectively useless to us i.e. we can't sync with it). I've also added in logic in `Peer::isUseless` to take the number of failed connection attempts into account when the last disconnect isn't critical.